### PR TITLE
Implemented .elf and .dol support in gamelist

### PR DESCRIPTION
**Pull request by aserna3**
This is to fix https://code.google.com/p/dolphin-emu/issues/detail?id=6985

~~This far from an ideal implementation or a clean one. I'm very open to improving this. The big problems were that while ISOs and WBFS files have plenty of metadata we can pull from and populate the list with info, dols and elfs have none of that. They're just an executable. So I had to do two things. One, implement a kinda hacky way to tell whether we're dealing with a dol/elf or a game. And two, implement hardcoded values on some UI fields that would throw a runtime error if their methods were run on an elf/dol due to not having that metadata at all.~~

@lioncash and @JosJuice helped me clean this up with some good ideas and now it's pretty awesome despite having no metadata.

At the end of the day, this allows dols and elfs to populate the netplay list and is a big win for Project M users and developers.

Screenshots because UI changes:

Image of main screen

Image of netplay

### CODE REVIEW

File: Source/Core/DolphinWX/ISOFile.h [hunk 1]
```diff
@@ -43,6 +43,8 @@ class GameListItem : NonCopyable
 	u64 GetVolumeSize() const {return m_VolumeSize;}
 	// 0 is the first disc, 1 is the second disc
 	u8 GetDiscNumber() const {return m_disc_number;}
+	bool GameListItem::isElfDol(const GameListItem& game);
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.h]: This can be a static function. It should also be CamelCased ``IsElfDol``

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 2]
```diff
@@ -282,4 +289,30 @@ const std::string GameListItem::GetWiiFSPath() const
...
+	{
+		if (game.GetName().substr(game.GetName().rfind(".")) == ".elf" ||
+			game.GetName().substr(game.GetName().rfind(".")) == ".dol")
+			return true;
+		else
+			return false;
+	}
+	return false;
+}
+bool GameListItem::isElfDol(const std::string& game)
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]:
```cpp
bool GameListItem::IsElfDol(const std::string& game)
{
    size_t pos = game.rfind('.');

    if (pos != std::string::npos)
    {
        return game.substr(pos) == ".elf" ||
               game.substr(pos) == ".dol";
    }

    return false;
}
```

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 3]
```diff
@@ -282,4 +289,30 @@ const std::string GameListItem::GetWiiFSPath() const
 
 	return ret;
 }
-
+bool GameListItem::isElfDol(const GameListItem& game)
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]:
```cpp
bool GameListItem::IsElfDol(const GameListItem& game)
{
    //This can probably be made cleaner - The way elf/dol support in the gamelist
    //is implemented isn't the greatest.
    size_t pos = game.GetName().rfind('.');

    if (pos != std::string::npos)
    {
        return (game.GetName().substr(pos) == ".elf" ||
                game.GetName().substr(pos) == ".dol")
    }

    return false;
}
```

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 4]
```diff
@@ -236,12 +240,15 @@ std::string GameListItem::GetName() const
...
 	if (name.empty())
 	{
 		// No usable name, return filename (better than nothing)
-		SplitPath(GetFileName(), nullptr, &name, nullptr);
+		SplitPath(GetFileName(), nullptr, &name, &ext);
+		return name + ext;
 	}
-	return name;
+	else
+		return name;
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: This needs to be braced

File: Source/Core/DolphinWX/ISOFile.h [hunk 5]
```diff
@@ -43,6 +43,8 @@ class GameListItem : NonCopyable
...
+	bool GameListItem::isElfDol(const std::string& game);
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.h]: This can be a static function.

**Comment by booto** [comment]
The lint fails are due to CR (\r) appended to the end of many lines, probably due to using some windows text editor.

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 3 above)
Comment (aserna3) [inline Source/Core/DolphinWX/ISOFile.cpp]: Wait, so does this replace both of my existing functions that I made?

Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: No, only the GameListItem one, since the string one is still used [here](#diff-2ebf3ad42e590cbc246f0a1f3e3e479aR151)

Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Actually, yes it can.

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 4 above)
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Actually, the else is unnecessary here.

Comment (aserna3) [inline Source/Core/DolphinWX/ISOFile.cpp]: oops, I missed that when I was cleaning up for the PR. thanks

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 6]
```diff
@@ -147,6 +147,10 @@ GameListItem::GameListItem(const std::string& _rFileName)
 		// default banner
 		m_Bitmap.LoadFile(StrToWxStr(File::GetThemeDir(SConfig::GetInstance().theme_name)) + "nobanner.png", wxBITMAP_TYPE_PNG);
 	}
+
+	if (isElfDol(_rFileName))
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: This should be an `else if` branch on the `if (LoadFromCache())` if statement at line 74.

e.g.

```cpp
if (LoadFromCache())
{
    // Code here
}
else if (IsElfOrDol())
{
}
else
{
    // Wii/GC code here.
}
```

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 3 above)
Comment (aserna3) [inline Source/Core/DolphinWX/ISOFile.cpp]: Are you sure? A string can't call GetName() on it like in your example above. Won't that fail?

Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Yes, you would be calling the one with no parameters.

**Comment by MaJoRoesch** [comment]
Screenshots please! I'm quite curious how they appear to a user.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 7]
```diff
@@ -482,7 +496,9 @@ void CGameListCtrl::ScanForISOs()
 	}
 	if (SConfig::GetInstance().m_ListWad)
 		Extensions.push_back("*.wad");
-
+	if (SConfig::GetInstance().m_ListElfDol)
+		Extensions.push_back("*.dol");
+		Extensions.push_back("*.elf");
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: This needs to be braced, it's still doing fallthrough.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 8]
```diff
@@ -282,4 +289,17 @@ const std::string GameListItem::GetWiiFSPath() const
...
+	const std::string name = GetName();
+	const size_t pos = name.rfind('.');
+
+	if (pos != std::string::npos)
+	{
+		return name.substr(pos) == ".elf" ||
+			name.substr(pos) == ".dol";
+	}
+	return false;
+}
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Need a newline at the end of the file.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 9]
```diff
@@ -147,6 +151,7 @@ GameListItem::GameListItem(const std::string& _rFileName)
 		// default banner
 		m_Bitmap.LoadFile(StrToWxStr(File::GetThemeDir(SConfig::GetInstance().theme_name)) + "nobanner.png", wxBITMAP_TYPE_PNG);
 	}
+
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Unnecessary newline

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 10]
```diff
@@ -282,4 +289,17 @@ const std::string GameListItem::GetWiiFSPath() const
...
+{
+	//This can probably be made cleaner - The way elf/dol support in the gamelist
+	//is implemented isn't the greatest.
+	const std::string name = GetName();
+	const size_t pos = name.rfind('.');
+
+	if (pos != std::string::npos)
+	{
+		return name.substr(pos) == ".elf" ||
+			name.substr(pos) == ".dol";
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: one space less

File: Source/Core/DolphinWX/ISOFile.h [hunk 11]
```diff
@@ -42,7 +42,8 @@ class GameListItem : NonCopyable
 	u64 GetFileSize() const {return m_FileSize;}
 	u64 GetVolumeSize() const {return m_VolumeSize;}
 	// 0 is the first disc, 1 is the second disc
-	u8 GetDiscNumber() const {return m_disc_number;}
+	u8 GetDiscNumber() const { return m_disc_number; }
+	bool GameListItem::IsElfOrDol() const;
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.h]: `GameListItem::` shouldn't be part of the name.

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 10 above)
Comment (aserna3) [inline Source/Core/DolphinWX/ISOFile.cpp]: I don't know what you mean by one space less, do I need a new line between functions? I noticed I was missing one and it looked weird so I just corrected that

**Comment by Rukario** [comment]
AWW YIS!

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 12]
```diff
@@ -283,3 +293,15 @@ const std::string GameListItem::GetWiiFSPath() const
...
+	const std::string name = GetName();
+	const size_t pos = name.rfind('.');
+
+	if (pos != std::string::npos)
+	{
+		return name.substr(pos) == ".elf" ||
+		       name.substr(pos) == ".dol";
+    }
+	return false;
+}
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Missing a newline here.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 13]
```diff
@@ -283,3 +293,15 @@ const std::string GameListItem::GetWiiFSPath() const
...
+bool GameListItem::IsElfOrDol() const
+{
+	const std::string name = GetName();
+	const size_t pos = name.rfind('.');
+
+	if (pos != std::string::npos)
+	{
+		return name.substr(pos) == ".elf" ||
+		       name.substr(pos) == ".dol";
+    }
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: Brace is off alignment.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 14]
```diff
@@ -236,10 +244,12 @@ std::string GameListItem::GetName() const
 {
 	bool wii = m_Platform != DiscIO::IVolume::GAMECUBE_DISC;
 	std::string name = GetName(SConfig::GetInstance().GetCurrentLanguage(wii));
+	std::string ext;
 	if (name.empty())
 	{
 		// No usable name, return filename (better than nothing)
-		SplitPath(GetFileName(), nullptr, &name, nullptr);
+		SplitPath(GetFileName(), nullptr, &name, &ext);
+		return name + ext;
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Won't this get rid of the `.` that the extension starts with? As an alternative, you could make a `wxFileName` object and call `GetFullName()` on it.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 15]
```diff
@@ -236,10 +244,12 @@ std::string GameListItem::GetName() const
 {
 	bool wii = m_Platform != DiscIO::IVolume::GAMECUBE_DISC;
 	std::string name = GetName(SConfig::GetInstance().GetCurrentLanguage(wii));
+	std::string ext;
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Should be declared inside the `if (name.empty())` scope

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 14 above)
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: It won't get rid of the period.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 16]
```diff
@@ -283,3 +294,17 @@ const std::string GameListItem::GetWiiFSPath() const
 	return ret;
 }
 
+bool GameListItem::IsElfOrDol() const
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: To truly make this more robust you would do:

```cpp
bool GameListItem::IsElfOrDoll() const
{
	const std::string name = GetName();
	const size_t pos = name.rfind('.');

	if (pos != std::string::npos)
	{
		std::string ext = name.substr(pos);
		std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
		
		return ext == ".elf" ||
		       ext == ".dol";
	}
	return false;
}
```

or convert it to a wxString and call ToLower on it.

**Comment by JosJuice** [comment]
I don't like that the code that calls ISOFile needs to check whether it's a DOL/ELF before it can use certain functions. This should be handled inside of ISOFile instead. For instance, `m_Country` can be set to COUNTRY_UNKNOWN in the constructor for DOLs/ELFs so that `GetCountry()` returns a good value.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 17]
```diff
@@ -129,7 +133,11 @@ GameListItem::GameListItem(const std::string& _rFileName)
 		ini.GetIfExists("EmuState", "EmulationIssues", &m_issues);
 	}
 
-	if (!m_pImage.empty())
+	if (IsElfOrDol())
+	{
+		m_Bitmap.LoadFile(StrToWxStr(File::GetThemeDir(SConfig::GetInstance().theme_name)) + "filebanner.png", wxBITMAP_TYPE_PNG);
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Why? DOLs/ELFs should have a separate platform icon, not a special banner. This feels odd from the user perspective and especially the code perspective.

File: Data/Sys/Themes/Boomy/README.txt [hunk 18]
```diff
@@ -1 +0,0 @@
-Boomy: Milosz Wlazlo [miloszwl@miloszwl.com]
```
Comment (JosJuice) [inline Data/Sys/Themes/Boomy/README.txt]: Why is this removed?

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 17 above)
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: You'll need to bring that up with MaJoR

**Comment by clshortfuse** [comment]
All Homebrew Channel homebrew normally include a 128x48 icon.png file that can use for a banner and a meta.xml which describe its contents.

Here's the layout: http://wiibrew.org/wiki/Homebrew_Channel#Adding_Text

File: Source/Core/DolphinWX/NetPlay/NetWindow.cpp [hunk 19]
```diff
@@ -75,6 +75,10 @@ static std::string BuildGameName(const GameListItem& game)
 	if (game.GetRevision() != 0)
 		return name + " (" + game.GetUniqueID() + ", Revision " + std::to_string((long long)game.GetRevision()) + ")";
 	else
+		if (game.IsElfOrDol())
+		{
+			return game.GetName();
+		}
 		return name + " (" + game.GetUniqueID() + ")";
```
Comment (waddlesplash) [inline Source/Core/DolphinWX/NetPlay/NetWindow.cpp]: Missing braces, or misleading indentation.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 20]
```diff
@@ -837,26 +854,35 @@ void CGameListCtrl::OnRightClick(wxMouseEvent& event)
...
+			if (!selected_iso->IsElfOrDol())
+				popupMenu.AppendCheckItem(IDM_SET_DEFAULT_ISO, _("Set as &default ISO"));
 
 			// First we have to decide a starting value when we append it
 			if (selected_iso->GetFileName() == SConfig::GetInstance().m_strDefaultISO)
 				popupMenu.FindItem(IDM_SET_DEFAULT_ISO)->Check();
 
-			popupMenu.AppendSeparator();
-			popupMenu.Append(IDM_DELETE_ISO, _("&Delete ISO..."));
+			if (!selected_iso->IsElfOrDol())
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: I don't see why you would not be able to delete a DOL or ELF here (except for the wording of the item, which can be changed to "Delete file..." or something more general).

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 21]
```diff
@@ -837,26 +854,35 @@ void CGameListCtrl::OnRightClick(wxMouseEvent& event)
...
+			{
+				popupMenu.AppendSeparator();
+				popupMenu.Append(IDM_DELETE_ISO, _("&Delete ISO..."));
+			}
+
 
-			if (selected_iso->GetPlatform() != DiscIO::IVolume::WII_WAD)
+			if (selected_iso->GetPlatform() != DiscIO::IVolume::WII_WAD && !selected_iso->IsElfOrDol())
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: This should probably be moved into ```ISOFile``` and provided as a property of the file itself: ```iso_file->SupportsCompression()``` or similar.

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 17 above)
Comment (BhaaLseN) [inline Source/Core/DolphinWX/ISOFile.cpp]: I'd agree tho. Instead of exposing ```IsElfOrDol``` and deciding in the UI, it would probably be nicer to simply return appropriate values within the given methods (such as ```GetCountry```, ```GetEmuState``` and even ```GetFileSize```). Especially since that way, other existing code wouldn't break if it called either of those on an ELF/DOL without checking for ```IsElfOrDol``` first.

File: Data/Sys/Themes/Boomy/README.txt (hunk 18 above)
Comment (aserna3) [inline Data/Sys/Themes/Boomy/README.txt]: Error. MaJoR didn't specify which READMEs to remove so I removed all of them. Fixing that in my next push.

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 21 above)
Comment (aserna3) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Are you talking about reworking that section so we don't do checks if it's a WAD, Wii, elf/dol, or GC game and instead just checks if it's compressible? I feel that's a bit out of the scope of this PR. I can make a different one later but I don't feel that blocks this one if I understand you correctly.

Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Since Options used to be exclusive and taken care of by this ```if```; yes, you have a point. But then again, multiple people have argued that exposing and using ```IsElfOrDol``` like that should be avoided if possible; so it is somewhat in-scope and somewhat out-of-scope.
And since you're touching this part anyways (both IsoFile and GameListCtrl here), you might aswell do it the proper way instead of hacking in something, then reworking it later.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 22]
```diff
@@ -283,3 +296,31 @@ const std::string GameListItem::GetWiiFSPath() const
...
+		return ext == ".elf" ||
+		       ext == ".dol";
+	}
+	return false;
+}
+
+bool GameListItem::SupportsCompression() const
+{
+	if (GetPlatform() == DiscIO::IVolume::GAMECUBE_DISC ||
+		GetPlatform() == DiscIO::IVolume::WII_DISC)
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: The second tab here needs to be replaced with spaces. Also, you can return the condition directly, removing the if statement.

**Comment by aserna3** [comment]
I'll remove the last elf/dol check in the UI once I get an icon from MaJoR.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 23]
```diff
@@ -416,18 +417,17 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
 
 	int disc_number = rISOFile.GetDiscNumber() + 1;
 	if (disc_number > 1 && name.Lower().find(wxString::Format("disc %i", disc_number)) == std::string::npos
-	                    && name.Lower().find(wxString::Format("disc%i", disc_number)) == std::string::npos)
+		&& name.Lower().find(wxString::Format("disc%i", disc_number)) == std::string::npos)
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: There's no reason to make this change.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 24]
```diff
@@ -416,18 +417,17 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
...
 	}
 
 	SetItem(_Index, COLUMN_TITLE, name, -1);
+
 	SetItem(_Index, COLUMN_MAKER, StrToWxStr(rISOFile.GetCompany()), -1);
 
-	// Emulation state
 	SetItemColumnImage(_Index, COLUMN_EMULATION_STATE, m_EmuStateImageIndex[rISOFile.GetEmuState()]);
 
-	// Country
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Same with these.

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 23 above)
Comment (aserna3) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: oops, that was my editor and I wasn't paying attention. I'll fix that.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 25]
```diff
@@ -129,7 +129,18 @@ GameListItem::GameListItem(const std::string& _rFileName)
...
 
-	if (!m_pImage.empty())
+	if (IsElfOrDol())
+	{
+		DiscIO::IVolume* pVolume = DiscIO::CreateVolumeFromFilename(_rFileName);
+
+		m_Bitmap.LoadFile(StrToWxStr(File::GetThemeDir(SConfig::GetInstance().theme_name)) + "filebanner.png", wxBITMAP_TYPE_PNG);
+		m_Country = DiscIO::IVolume::COUNTRY_UNKNOWN;
+		m_emu_state = 0;
+		m_disc_number = 0;
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: You could move the initialization of `m_disc_number` and `m_Country` to the initializer list. `m_emu_state` is already in the initializer list and is thus unnecessary here.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 26]
```diff
@@ -837,26 +842,40 @@ void CGameListCtrl::OnRightClick(wxMouseEvent& event)
...
+			}
+			else
+			{
+				popupMenu.AppendSeparator();
+				popupMenu.Append(IDM_DELETE_ISO, _("&Delete File..."));
+			}
+
 
-			if (selected_iso->GetPlatform() != DiscIO::IVolume::WII_WAD)
+			if (selected_iso->SupportsCompression())
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: I know someone told you to add the `SupportsCompression function, but it makes the ifs and elses here rather unclear :/
That the final else means "only WADs" is not at all obvious unless you know that `SupportsCompression` means Wii and GC discs and that those combined with DOLs and ELFs are the only existing platforms.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 27]
```diff
@@ -837,26 +842,40 @@ void CGameListCtrl::OnRightClick(wxMouseEvent& event)
...
 			// First we have to decide a starting value when we append it
 			if (selected_iso->GetFileName() == SConfig::GetInstance().m_strDefaultISO)
 				popupMenu.FindItem(IDM_SET_DEFAULT_ISO)->Check();
 
-			popupMenu.AppendSeparator();
-			popupMenu.Append(IDM_DELETE_ISO, _("&Delete ISO..."));
+			if (!selected_iso->IsElfOrDol())
+			{
+				popupMenu.AppendSeparator();
+				popupMenu.Append(IDM_DELETE_ISO, _("&Delete ISO..."));
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: You should remove this and just use the string "&Delete File..." for everything. ISO isn't the correct term when you're dealing with a WAD, for instance.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 28]
```diff
@@ -416,18 +417,17 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
 
 	int disc_number = rISOFile.GetDiscNumber() + 1;
 	if (disc_number > 1 && name.Lower().find(wxString::Format("disc %i", disc_number)) == std::string::npos
-	                    && name.Lower().find(wxString::Format("disc%i", disc_number)) == std::string::npos)
+		                && name.Lower().find(wxString::Format("disc%i", disc_number)) == std::string::npos)
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Still wrong.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 29]
```diff
@@ -416,18 +417,17 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
...
 	}
 
 	SetItem(_Index, COLUMN_TITLE, name, -1);
+
 	SetItem(_Index, COLUMN_MAKER, StrToWxStr(rISOFile.GetCompany()), -1);
 
-	// Emulation state
 	SetItemColumnImage(_Index, COLUMN_EMULATION_STATE, m_EmuStateImageIndex[rISOFile.GetEmuState()]);
 
-	// Country
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: It's a bit weird to remove these comments while keeping the ones below.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 30]
```diff
@@ -864,6 +874,10 @@ void CGameListCtrl::OnRightClick(wxMouseEvent& event)
 				         selected_iso->GetFileName().substr(selected_iso->GetFileName().find_last_of(".")) != ".wbfs")
 					popupMenu.Append(IDM_COMPRESS_ISO, _("Compress ISO..."));
 			}
+			else if (selected_iso->IsElfOrDol())
+			{
+
+			}
 			else
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Instead of having an empty else-if, you could replace the final `else` with `else if (!selected_iso->IsElfOrDol()))`

File: Source/Core/DolphinWX/NetPlay/NetWindow.cpp [hunk 31]
```diff
@@ -75,7 +75,13 @@ static std::string BuildGameName(const GameListItem& game)
 	if (game.GetRevision() != 0)
 		return name + " (" + game.GetUniqueID() + ", Revision " + std::to_string((long long)game.GetRevision()) + ")";
 	else
+	{
+		if (game.IsElfOrDol())
```
Comment (JosJuice) [inline Source/Core/DolphinWX/NetPlay/NetWindow.cpp]: A better way to do this is to check whether the value of `GetUniqueID()` is empty.

File: Source/Core/DolphinWX/NetPlay/NetWindow.cpp [hunk 32]
```diff
@@ -75,7 +75,13 @@ static std::string BuildGameName(const GameListItem& game)
 	if (game.GetRevision() != 0)
 		return name + " (" + game.GetUniqueID() + ", Revision " + std::to_string((long long)game.GetRevision()) + ")";
 	else
+	{
+		if (game.GetUniqueID() == "")
```
Comment (JosJuice) [inline Source/Core/DolphinWX/NetPlay/NetWindow.cpp]: `game.GetUniqueID().empty()`

**Comment by bb010g** [comment]
:+1: for pulling from Homebrew Channel metadata

**Comment by aserna3** [comment]
@bb010g doesn't work like that all of the time. You're assuming the xml file will exist

**Comment by JosJuice** [comment]
If meta.xml doesn't exist, just do what the PR is doing now.

**Comment by aserna3** [comment]
The banner used in this PR right now won't work because it won't translate. Would this be PR be fine without a platform icon for the time being and falling back onto nobanner.png? Thinking of a platform icon for this that doesn't run into translation issues is hard.

**Comment by JosJuice** [comment]
What about a generic file icon as a platform icon?

**Comment by aserna3** [comment]
I don't think @MaJoRoesch was too keen about that. I'll ask her again.

**Comment by aserna3** [comment]
Another idea which I personally don't like at all but would solve the icon issue is refactoring the code to differentiate between elfs and dols. I don't like that idea because it really doesn't matter to the user which one it is as it's all handled in the UI the same way and effectively means the same thing to the user. It wouldn't be hard to do, I could do it, but eh. I don't like it.

**Comment by JosJuice** [comment]
If we can't come up with a good icon, I'd like to see the icon implemented in the code, but with an blank image file. It fixes some of the code cleanness that I commented on, and it makes the icon simple to add at a later time.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 33]
```diff
@@ -422,9 +423,9 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
 	}
 
 	SetItem(_Index, COLUMN_TITLE, name, -1);
+
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: unnecessary newline

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 34]
```diff
@@ -70,6 +70,8 @@ GameListItem::GameListItem(const std::string& _rFileName)
 	, m_BlobCompressed(false)
 	, m_ImageWidth(0)
 	, m_ImageHeight(0)
+	, m_disc_number(0)
+	, m_Country(DiscIO::IVolume::COUNTRY_UNKNOWN)
```
Comment (lioncash) [inline Source/Core/DolphinWX/ISOFile.cpp]: m_Country needs to go above m_Valid, otherwise it'll cause warnings on Linux/OSX

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 35]
```diff
@@ -283,3 +294,18 @@ const std::string GameListItem::GetWiiFSPath() const
...
+	const std::string name = GetName();
+	const size_t pos = name.rfind('.');
+
+	if (pos != std::string::npos)
+	{
+		std::string ext = name.substr(pos);
+		std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
+
+		return ext == ".elf" ||
+		       ext == ".dol";
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: The second tab here is wrong. I suggest just putting this on one line, it's very short.

File: Source/Core/DolphinWX/FrameTools.cpp [hunk 36]
```diff
@@ -308,6 +308,8 @@ wxMenuBar* CFrame::CreateMenu()
 	platformMenu->Check(IDM_LIST_GC, SConfig::GetInstance().m_ListGC);
 	platformMenu->AppendCheckItem(IDM_LIST_WAD, _("Show Wad"));
 	platformMenu->Check(IDM_LIST_WAD, SConfig::GetInstance().m_ListWad);
+	platformMenu->AppendCheckItem(IDM_LIST_ELFDOL, _("Show ELFs/DOLs"));
```
Comment (JosJuice) [inline Source/Core/DolphinWX/FrameTools.cpp]: Inconsistent with "Show Wad". Either change this to "Show ELF/DOL" or change "Show Wad" to "Show WADs".

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 37]
```diff
@@ -129,7 +131,14 @@ GameListItem::GameListItem(const std::string& _rFileName)
 		ini.GetIfExists("EmuState", "EmulationIssues", &m_issues);
 	}
 
-	if (!m_pImage.empty())
+	if (IsElfOrDol())
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Check `!IsValid()` or `!pVolume` before checking `IsElfOrDol()`, otherwise weird things will happen if someone has a valid volume that has the file name extension .dol or .elf.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 38]
```diff
@@ -129,7 +131,14 @@ GameListItem::GameListItem(const std::string& _rFileName)
...
+	{
+		m_Bitmap.LoadFile(StrToWxStr(File::GetThemeDir(SConfig::GetInstance().theme_name)) + "nobanner.png", wxBITMAP_TYPE_PNG);
+		m_Valid = true;
+		m_FileSize = File::GetSize(_rFileName);
+		m_Platform = DiscIO::IVolume::ELF_DOL;
+	}
+	else if (!m_pImage.empty())
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Why `else if`? This could just be a separate `if` statement, then the line above for loading `nobanner.png` will be unnecessary.

File: Source/Core/DolphinWX/FrameTools.cpp [hunk 39]
```diff
@@ -308,6 +308,8 @@ wxMenuBar* CFrame::CreateMenu()
 	platformMenu->Check(IDM_LIST_GC, SConfig::GetInstance().m_ListGC);
 	platformMenu->AppendCheckItem(IDM_LIST_WAD, _("Show Wad"));
 	platformMenu->Check(IDM_LIST_WAD, SConfig::GetInstance().m_ListWad);
+	platformMenu->AppendCheckItem(IDM_LIST_ELFDOL, _("Show Elfs/Dols"));
```
Comment (JosJuice) [inline Source/Core/DolphinWX/FrameTools.cpp]: I meant that the pluralization needs to be consistent. Making the capitalization consistent is also nice, though.

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 40]
```diff
@@ -129,6 +131,12 @@ GameListItem::GameListItem(const std::string& _rFileName)
 		ini.GetIfExists("EmuState", "EmulationIssues", &m_issues);
 	}
 
+	if (!IsValid && IsElfOrDol())
+	{
+		m_Valid = true;
+		m_FileSize = File::GetSize(_rFileName);
+		m_Platform = DiscIO::IVolume::ELF_DOL;
+	}
 	if (!m_pImage.empty())
```
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Having a newline above this line would be nice.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 41]
```diff
@@ -424,7 +425,6 @@ void CGameListCtrl::InsertItemInReportView(long _Index)
 	SetItem(_Index, COLUMN_TITLE, name, -1);
 	SetItem(_Index, COLUMN_MAKER, StrToWxStr(rISOFile.GetCompany()), -1);
 
-	// Emulation state
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: This comment shouldn't be removed.

**Comment by JosJuice** [comment]
Other than that little thing, LGTM.

It would be good to have the platform images unified in the future, probably by moving them all to PNGs in themes.

**Comment by Rukario** [comment]
I was hoping you could include this UNK flag I made here http://i.imgur.com/1ddlsst.png (it's @2x
 btw)

**Comment by aserna3** [comment]
ping @MaJoRoesch in IRC or the forums about it.

**Comment by MaJoRoesch** [comment]
@Rukario I've already redone all of the flags, and they have been applied to Qt, but they haven't been applied to WX yet because the flags are XMP files...

**Comment by shuffle2** [comment]
re: @booto (but really for @aserna3 )...there is a git setting which you should be using, which should transparently fix any future possible newline messups: https://help.github.com/articles/dealing-with-line-endings/  
If you're using VS, you can see/set the setting here:  
here

**Comment by Rukario** [comment]
@MaJoRoesch Now that made sense... Other days I've added new UNK flag to the \resources\ folder and wasn't able to understand why it's not loaded in Dolphin Emulator. And, oh, when I was going to check Flag_Unknown.png in \resource\ folder again recently and just noticed it's updated! It's X instead of UNK, hmm...
